spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java

```diff
@@ -333,6 +352,14 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
 					error = ((ErrorMessage) reply).getPayload();
 				}
 			}
+			if (reply == null && this.errorOnTimeout) {
+				if (object instanceof Message) {
+					error = new MessageTimeoutException((Message<?>) object, "No reply received within timeout");
+				}
+				else {
+					error = new MessageTimeoutException("No reply received within timeout");
+				}
```

Comment:
After the private internal discussion decided to leave as is, because only HTTP Inbound is the user of the feature right now.
Will be revised in the `4.3` version.